File: src/alerts.rs
//! Alert models, filtering, formatting, and StockCharts timestamp parsing.

use chrono::{DateTime, LocalResult, NaiveDateTime, TimeZone};
use chrono_tz::Tz;
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;
use tracing::warn;

use crate::{Error, Result};

/// Time zone used by StockCharts alert timestamps.
pub const STOCKCHARTS_TIME_ZONE: Tz = chrono_tz::America::New_York;

const NO_ALERTS_PLACEHOLDER: &str = "There are no alerts today";

/// StockCharts alert payload fields used by the bot.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct Alert {
    /// Human-readable alert text.
    #[serde(deserialize_with = "deserialize_trimmed_string")]
    pub alert: String,
    /// Whether the alert is bearish. StockCharts uses the string "yes" for bearish alerts.
    #[serde(
        default = "default_bearish",
        deserialize_with = "deserialize_trimmed_string"
    )]
    pub bearish: String,
    /// Timestamp string from StockCharts.
    #[serde(deserialize_with = "deserialize_trimmed_string")]
    pub lastfired: String,
    /// Stock or index symbol. Missing values match the Python default.
    #[serde(
        default = "default_symbol",
        deserialize_with = "deserialize_trimmed_string"
    )]
    pub symbol: String,
}

impl Alert {
    /// Parse an alert from an arbitrary JSON value, ignoring unknown fields.
    ///
    /// # Errors
    ///
    /// Returns an error when required alert fields are missing or invalid.
    pub fn from_value(value: Value) -> Result<Self> {
        serde_json::from_value(value).map_err(|error| Error::AlertPayload(error.to_string()))
    }
}

/// Return alerts that are valid, not placeholder rows, and newer than `previous_run`.
#[must_use]
pub fn new_alerts_since(alerts: &[Value], previous_run: DateTime<Tz>) -> Vec<Alert> {
    let parsed_alerts = filter_alerts(alerts)
        .into_iter()
        .filter_map(|alert| match parse_stockcharts_time(&alert.lastfired) {
            Ok(fired_at) if fired_at > previous_run => Some((alert, fired_at)),
            Ok(_) => None,
            Err(error) => {
                warn!(symbol = %alert.symbol, %error, "failed to parse StockCharts alert timestamp");
                None
            }
        })
        .collect::<Vec<_>>();

    parsed_alerts
        .iter()
        .filter(|(alert, fired_at)| {
            parsed_alerts
                .iter()
                .filter(|(other, _)| other.symbol == alert.symbol)
                .all(|(_, other_fired_at)| other_fired_at <= fired_at)
        })
        .map(|(alert, _)| alert.clone())
        .collect()
}

/// Return valid, sendable alerts from StockCharts response payloads.
#[must_use]
pub fn filter_alerts(alerts: &[Value]) -> Vec<Alert> {
    alerts
        .iter()
        .filter_map(|value| match Alert::from_value(value.clone()) {
            Ok(alert) if alert.alert != NO_ALERTS_PLACEHOLDER => Some(alert),
            Ok(_) => None,
            Err(error) => {
                warn!(%error, "skipping malformed StockCharts alert payload");
                None
            }
        })
        .collect()
}

/// Parse a StockCharts timestamp in the America/New_York timezone.
///
/// # Errors
///
/// Returns an error when the timestamp format is unsupported or maps to no local time.
pub fn parse_stockcharts_time(value: &str) -> Result<DateTime<Tz>> {
    let cleaned = value.trim().trim_end_matches(" ET").trim();
    let naive = parse_naive_stockcharts_time(cleaned)?;

    match STOCKCHARTS_TIME_ZONE.from_local_datetime(&naive) {
        LocalResult::Single(datetime) => Ok(datetime),
        LocalResult::Ambiguous(earliest, _) => Ok(earliest),
        LocalResult::None => Err(Error::TimeParse(format!(
            "{value:?} is not a valid Eastern time"
        ))),
    }
}

fn parse_naive_stockcharts_time(value: &str) -> Result<NaiveDateTime> {
    const FORMATS: &[&str] = &["%e %b %Y, %-I:%M%P", "%e %b %Y, %-I:%M %p"];

    FORMATS
        .iter()
        .find_map(|format| NaiveDateTime::parse_from_str(value, format).ok())
        .ok_or_else(|| Error::TimeParse(format!("unsupported StockCharts timestamp {value:?}")))
}

fn default_bearish() -> String {
    "no".to_string()
}

fn default_symbol() -> String {
    "UNKNOWN".to_string()
}

fn deserialize_trimmed_string<'de, D>(deserializer: D) -> std::result::Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    String::deserialize(deserializer).map(|value| value.trim().to_string())
}

#[cfg(test)]
mod tests {
    use chrono::TimeZone;
    use serde_json::json;

    use super::{
        Alert, STOCKCHARTS_TIME_ZONE, filter_alerts, new_alerts_since, parse_stockcharts_time,
    };

    fn sample_alerts() -> Vec<serde_json::Value> {
        vec![
            json!({
                "alert": "There are no alerts today",
                "newalert": "yes",
                "bearish": "",
                "lastfired": "1 Aug 2024, 8:11 AM ET"
            }),
            json!({
                "symbol": "$BPSPX",
                "alertpaused": "no",
                "bearish": "no",
                "notes": "",
                "alert": "S&P 500 Bullish Percent Index crosses above 70",
                "lastfired": "31 Jul 2024, 2:31pm",
                "newalert": "yes",
                "type": "a",
                "recid": "701"
            }),
            json!({
                "symbol": "$BPINFO",
                "alertpaused": "no",
                "bearish": "yes",
                "notes": "",
                "alert": "Technology Sector Bullish Percent Index crosses below 50",
                "lastfired": "31 Jul 2024, 12:55pm",
                "newalert": "yes",
                "type": "a",
                "recid": "1739"
            }),
            json!({
                "symbol": "$INDU",
                "alertpaused": "no",
                "bearish": "no",
                "notes": "",
                "alert": "Dow crosses above 41000",
                "lastfired": "31 Jul 2024, 12:33pm",
                "newalert": "yes",
                "type": "a",
                "recid": "452083"
            }),
            json!({
                "symbol": "$COMPQ",
                "alertpaused": "no",
                "bearish": "yes",
                "notes": "",
                "alert": "Nasdaq crosses below 17200",
                "lastfired": "31 Jul 2024, 11:47am",
                "newalert": "yes",
                "type": "a",
                "recid": "450121"
            }),
            json!({
                "symbol": "$COMPQ",
                "alertpaused": "no",
                "bearish": "yes",
                "notes": "",
                "alert": "Nasdaq crosses below 17300",
                "lastfired": "31 Jul 2024, 11:47am",
                "newalert": "yes",
                "type": "a",
                "recid": "450208"
            }),
        ]
    }

    #[test]
    fn alert_defaults_match_python_model() {
        let alert = Alert::from_value(json!({
            "alert": " Test alert ",
            "lastfired": " 31 Jul 2024, 12:33pm ",
            "ignored": "field"
        }))
        .expect("alert should deserialize");

        assert_eq!(alert.alert, "Test alert");
        assert_eq!(alert.lastfired, "31 Jul 2024, 12:33pm");
        assert_eq!(alert.bearish, "no");
        assert_eq!(alert.symbol, "UNKNOWN");
    }

    #[test]
    fn filter_alerts_skips_placeholder_and_malformed_rows() {
        let mut alerts = sample_alerts();
        alerts.push(json!({"alert": "missing lastfired"}));

        let filtered = filter_alerts(&alerts);

        assert_eq!(filtered.len(), 5);
        assert_eq!(filtered[0].symbol, "$BPSPX");
    }

    #[test]
    fn new_alerts_since_matches_existing_python_filter() {
        let previous_run = STOCKCHARTS_TIME_ZONE
            .with_ymd_and_hms(2024, 7, 31, 12, 0, 0)
            .single()
            .expect("valid timestamp");

        let alerts = new_alerts_since(&sample_alerts(), previous_run);

        assert_eq!(alerts.len(), 3);
        assert_eq!(
            alerts
                .iter()
                .map(|alert| alert.symbol.as_str())
                .collect::<Vec<_>>(),
            vec!["$BPSPX", "$BPINFO", "$INDU"]
        );
    }

    #[test]
    fn new_alerts_since_keeps_latest_alerts_per_symbol() {
        let previous_run = STOCKCHARTS_TIME_ZONE
            .with_ymd_and_hms(2026, 6, 8, 9, 29, 0)
            .single()
            .expect("valid timestamp");
        let alerts = vec![
            json!({
                "symbol": "$COMPQ",
                "bearish": "yes",
                "alert": "Nasdaq crosses below 25800",
                "lastfired": "8 Jun 2026, 9:30am"
            }),
            json!({
                "symbol": "$COMPQ",
                "bearish": "no",
                "alert": "Nasdaq crosses above 25800",
                "lastfired": "8 Jun 2026, 9:33am"
            }),
            json!({
                "symbol": "$COMPQ",
                "bearish": "no",
                "alert": "Nasdaq crosses above 25900",
                "lastfired": "8 Jun 2026, 9:33am"
            }),
            json!({
                "symbol": "$GOLD",
                "bearish": "yes",
                "alert": "Gold crosses below 4400",
                "lastfired": "8 Jun 2026, 9:30am"
            }),
        ];

        let alerts = new_alerts_since(&alerts, previous_run);

        assert_eq!(alerts.len(), 3);
        assert_eq!(
            alerts
                .iter()
                .map(|alert| alert.alert.as_str())
                .collect::<Vec<_>>(),
            vec![
                "Nasdaq crosses above 25800",
                "Nasdaq crosses above 25900",
                "Gold crosses below 4400"
            ]
        );
    }

    #[test]
    fn parse_stockcharts_time_uses_eastern_time() {
        let parsed = parse_stockcharts_time("31 Jul 2024, 2:31pm").expect("timestamp should parse");

        assert_eq!(parsed.timezone(), STOCKCHARTS_TIME_ZONE);
        assert_eq!(parsed.to_rfc3339(), "2024-07-31T14:31:00-04:00");
    }

    #[test]
    fn parse_stockcharts_time_accepts_uppercase_am_pm_and_et_suffix() {
        let parsed =
            parse_stockcharts_time("1 Aug 2024, 8:11 AM ET").expect("timestamp should parse");

        assert_eq!(parsed.to_rfc3339(), "2024-08-01T08:11:00-04:00");
    }

    #[test]
    fn parse_stockcharts_time_handles_dst_fall_ambiguity_deterministically() {
        let parsed =
            parse_stockcharts_time("3 Nov 2024, 1:30am").expect("ambiguous timestamp should parse");

        assert_eq!(parsed.to_rfc3339(), "2024-11-03T01:30:00-04:00");
    }
}

File: internal/alerts/alerts.go
// Package alerts provides alert models, filtering, and StockCharts timestamp parsing.
package alerts

import (
	"encoding/json"
	"log/slog"
	"strings"
	"time"
	// Blank-imported to embed the IANA timezone database in the binary,
	// so timezone lookups work even without system tzdata (e.g. in minimal container images).
	_ "time/tzdata"

	"github.com/major/stockchartsalerts/internal/xerrors"
)

// Alert represents a StockCharts alert payload.
type Alert struct {
	// Alert is the human-readable alert text.
	Alert string
	// Bearish indicates whether the alert is bearish ("yes" or "no").
	Bearish string
	// LastFired is the timestamp string from StockCharts.
	LastFired string
	// Symbol is the stock or index symbol.
	Symbol string
}

const (
	// NoAlertsPlaceholder is the placeholder text for days with no alerts.
	NoAlertsPlaceholder = "There are no alerts today"
	// StockChartsTimeZoneName is the IANA timezone name used by StockCharts.
	StockChartsTimeZoneName = "America/New_York"
)

// stockChartsTimeZone is the cached America/New_York timezone.
// Initialized once at package init time.
var stockChartsTimeZone *time.Location

func init() {
	var err error
	stockChartsTimeZone, err = time.LoadLocation(StockChartsTimeZoneName)
	if err != nil {
		// This should never happen with the embedded tzdata, but fail loudly if it does.
		panic("failed to load America/New_York timezone: " + err.Error())
	}
}

// StockChartsTimeZone returns the cached America/New_York timezone.
// This timezone is loaded once at package initialization and is guaranteed to be valid.
func StockChartsTimeZone() *time.Location {
	return stockChartsTimeZone
}

// UnmarshalJSON unmarshals an Alert from JSON, applying field defaults (Bearish
// defaults to "no", Symbol defaults to "UNKNOWN" when absent) and trimming
// surrounding whitespace from every string field.
func (a *Alert) UnmarshalJSON(data []byte) error {
	// Use a raw struct with pointer fields to detect missing values.
	type rawAlert struct {
		Alert     *string `json:"alert"`
		Bearish   *string `json:"bearish"`
		LastFired *string `json:"lastfired"`
		Symbol    *string `json:"symbol"`
	}

	var raw rawAlert
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}

	// Apply defaults and trim all fields.
	a.Alert = trimString(raw.Alert, "")
	a.Bearish = trimString(raw.Bearish, "no")
	a.LastFired = trimString(raw.LastFired, "")
	a.Symbol = trimString(raw.Symbol, "UNKNOWN")

	return nil
}

// trimString returns the trimmed value if non-nil and non-empty, otherwise the default.
func trimString(value *string, defaultValue string) string {
	if value == nil {
		return defaultValue
	}
	trimmed := strings.TrimSpace(*value)
	if trimmed == "" {
		return defaultValue
	}
	return trimmed
}

// FilterAlerts returns valid, sendable alerts from StockCharts response payloads.
// It drops the "There are no alerts today" placeholder rows and malformed/unparseable rows,
// logging and skipping rather than erroring.
func FilterAlerts(alerts []json.RawMessage) []Alert {
	var result []Alert
	for _, rawMsg := range alerts {
		var alert Alert
		if err := json.Unmarshal(rawMsg, &alert); err != nil {
			slog.Warn("skipping malformed StockCharts alert payload", "error", err)
			continue
		}
		if alert.Alert == NoAlertsPlaceholder {
			continue
		}
		result = append(result, alert)
	}
	return result
}

// NewAlertsSince returns alerts that are valid, not placeholder rows, and newer than previousRun.
// For alerts sharing the same symbol, it keeps only the one(s) at the latest fired-at timestamp
// (ties are kept).
func NewAlertsSince(alerts []json.RawMessage, previousRun time.Time) []Alert {
	filtered := FilterAlerts(alerts)

	// Parse timestamps and filter by previousRun.
	type alertWithTime struct {
		alert   Alert
		firedAt time.Time
	}
	var parsed []alertWithTime
	for _, alert := range filtered {
		firedAt, err := ParseStockChartsTime(alert.LastFired)
		if err != nil {
			slog.Warn("failed to parse StockCharts alert timestamp", "symbol", alert.Symbol, "error", err)
			continue
		}
		if firedAt.After(previousRun) {
			parsed = append(parsed, alertWithTime{alert, firedAt})
		}
	}

	// Keep only the latest-fired alert(s) per symbol.
	var result []Alert
	for _, candidate := range parsed {
		isLatest := true
		for _, other := range parsed {
			if other.alert.Symbol == candidate.alert.Symbol && other.firedAt.After(candidate.firedAt) {
				isLatest = false
				break
			}
		}
		if isLatest {
			result = append(result, candidate.alert)
		}
	}

	return result
}

// ParseStockChartsTime parses a StockCharts timestamp in the America/New_York timezone.
// It strips trailing " ET" (case-sensitive), then parses against known formats.
// For DST fall-back ambiguity, it picks the earliest instant.
func ParseStockChartsTime(value string) (time.Time, error) {
	// Strip trailing " ET" and trim.
	cleaned := strings.TrimSpace(value)
	cleaned = strings.TrimSuffix(cleaned, " ET")
	cleaned = strings.TrimSpace(cleaned)

	// Try all known formats.
	// StockCharts uses two formats:
	// 1. "31 Jul 2024, 2:31pm" (lowercase am/pm, no space)
	// 2. "1 Aug 2024, 8:11 AM" (uppercase AM/PM, with space)
	// Go's time.Parse is case-sensitive for AM/PM, so we need both variants.
	formats := []string{
		"2 Jan 2006, 3:04pm",  // lowercase am/pm, no space
		"2 Jan 2006, 3:04 PM", // uppercase AM/PM, with space
	}

	var parsedTime time.Time
	for _, format := range formats {
		t, err := time.ParseInLocation(format, cleaned, stockChartsTimeZone)
		if err == nil {
			parsedTime = t
			return parsedTime, nil
		}
	}

	// If no format matched, return an error.
	return time.Time{}, xerrors.TimeParse("unsupported StockCharts timestamp: " + value)
}
